principles.rst

```diff
@@ -36,55 +36,273 @@ but lacks the unarticulated principles, which will be added over time.
+the ``.`` vs ``->`` in a ``forall`` or the presence or absence of a ``@``) should desugar
+to the same Core program.
+
+Currently, the design for `#281`_ (along with the design for ``-XTypeApplications``)
+violates the `Visibility Orthogonality Principle`_, because the visibility marker ``@`` also affects the difference between
+term-syntax and type-syntax. However, given the `Syntactic Unification Principle`_, we strive to uphold the `Visibility Orthogonality Principle`_ when
+there is an absence of punning.
+
+*Motivation:* Visibility should be just that: a superficial property that describes
+(only) whether an argument is visible in the user-written source code.
+
+Name resolution and scoping
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Lexical Scoping Principle (LSP)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _`Lexical Scoping Principle`:
+
+**Principle**:
+
+a. For every appearance of
+an identifier, it is possible to determine whether that appearance is a *binding site*
+or an *occurrence*, without involving the type system.
+
+b. For every *occurrence* of an
 identifier, it is possible to uniquely identify its *binding site*, without
 involving the type system.
 
-The LSP_ is true today, though Template Haskell splices may need to be run before
-completing name resolution (and running those splices requires type-checking them).
+The `Lexical Scoping Principle`_ is true today, with two complications:
+
+1. Template Haskell splices may need to be run before completing name resolution (and running those splices requires type-checking them).
+
+2. The `deprecated mechanism <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/exts/duplicate_record_fields.html#selector-functions>`_ for disambiguating duplicate record fields violates the `Lexical Scoping Principle`_ by requiring the type system.
+
+*Motivation:* These principles mean that we can understand the binding
+structure of a program without relying on type inference, important both for the
+implementation of GHC and the sanity of programmers.
+
+(a) from `#448`_; (b) from `#378`_.
+
+Local Lexical Scoping Principle (LLSP)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _`Local Lexical Scoping Principle`:
+
+**Principle**: For every appearance of an identifier, it is possible to determine
+whether that appearance is a *binding site* or an *occurrence*, without looking to see what identifiers are
+already in scope.
+
+This is a stronger version of the `Lexical Scoping Principle`_, part (a), that forbids even knowing what is in scope.
+
+The `Local Lexical Scoping Principle`_ is not true today, because of pattern signatures. If we have ``f (x :: Maybe a)``, the ``a``
+is an occurrence if ``a`` is already in scope, and it is a binding site otherwise.
+
+Along similar lines (though not strictly affected by the current phrasing of the `Local Lexical Scoping Principle`_), the
+type signature ``f :: a -> a`` might implicitly include ``forall a.`` or not, depending on whether ``a`` is in scope.
```

Comment:
I don't see how `f :: a -> a` is any different from `f (x :: Maybe a) (y :: a)` wrt the LLSP. 

In my view, none of the `a`s in either fragment are binding sites, they are all occurrences. We have a special rule for type variables that inserts an implicit, invisible binder at the top-level for free type variables.

Under this reading, I think LLSP becomes **weaker** than LSP (well, the original LSP, now LSP (b)). It's strictly a syntactic property, and I believe it is true of Haskell today.

Reply:
That's my feeling too, but there is more desire to grandfather in their behavior. See the alternatives and earlier discussion about this.

Reply:
In `f (x :: Maybe a) (y :: a) = ...`, if the first `a` is an occurrence, where is its binding site? We have no syntax to write one. The later sections of this proposal (involving `let`) allow us to write the binding site. Then we can properly say that the `a` is an occurrence, with its binding site in an implicit `let` block.

Reply:
> In `f (x :: Maybe a) (y :: a) = ...`, if the first `a` is an occurrence, where is its binding site?

There is no explicit binding site, which is weird! My point is that this is no different than `f :: a -> a` not having an explicit binding site for `a`. 

I agree we should have syntax to write an explicit binding site for both the term- and type-level syntaxes, and this principle would be right to call out the lack of an explicit binding syntax in the term-level syntax as a gap. I just wouldn't call the first `a` a binding in the meantime, it just adds confusion in my opinion.

Reply:
Yeah `-XNoPatternBinds` might be disergonomic to some, but it gives us conceptual crystal clarity. We are *so close* to finally making things make complete sense, I rather go all the way to have a solid foundation and then turn around and ask what we can do to make things ergonomic.